flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProviderImpl.java

```diff
@@ -0,0 +1,347 @@
+         * LOGICAL_AVRO_TYPES_TO_PROTO: Map containing mapping from Primitive Avro Schema Type to FieldDescriptorProto.
+         */
+        LOGICAL_AVRO_TYPES_TO_PROTO = new HashMap<>();
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.date().getName(), FieldDescriptorProto.Type.TYPE_INT32);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.decimal(1).getName(), FieldDescriptorProto.Type.TYPE_BYTES);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.timestampMicros().getName(), FieldDescriptorProto.Type.TYPE_INT64);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.timestampMillis().getName(), FieldDescriptorProto.Type.TYPE_INT64);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.uuid().getName(), FieldDescriptorProto.Type.TYPE_STRING);
+        // These are newly added.
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.timeMillis().getName(), FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.timeMicros().getName(), FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.localTimestampMillis().getName(),
+                FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put(
+                LogicalTypes.localTimestampMicros().getName(),
+                FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put("geography_wkt", FieldDescriptorProto.Type.TYPE_STRING);
+        LOGICAL_AVRO_TYPES_TO_PROTO.put("Json", FieldDescriptorProto.Type.TYPE_STRING);
+    }
+
+    // --------------- Obtain AvroSchema from TableSchema -----------------
+    /**
+     * Function to convert TableSchema to Avro Schema.
+     *
+     * @param tableSchema A {@link TableSchema} object to cast to {@link Schema}.
+     * @return Converted Avro Schema
+     */
+    private static Schema getAvroSchema(TableSchema tableSchema) {
+        return SchemaTransform.toGenericAvroSchema("root", tableSchema.getFields());
+    }
+
+    // --------------- Obtain Descriptor Proto from Avro Schema  ---------------
+    /**
+     * Obtains a Descriptor Proto by obtaining Descriptor Proto field by field.
+     *
+     * <p>Iterates over Avro Schema to obtain FieldDescriptorProto for it.
+     *
+     * @param schema Avro Schema, for which descriptor is needed.
+     * @return DescriptorProto describing the Schema.
+     */
+    private static DescriptorProto getDescriptorSchemaFromAvroSchema(Schema schema) {
+        // Iterate over each table field and add them to the schema.
+        Preconditions.checkState(!schema.getFields().isEmpty());
+
+        DescriptorProto.Builder descriptorBuilder = DescriptorProto.newBuilder();
+        // Create a unique name for the descriptor ('-' characters cannot be used).
+        // Replace with "_" and prepend "D".
```

Comment:
"Replace with "_" and prepend "D"."

you can remove this part of the comment since it's an implementation detail of bqSanitizedRandomUUIDForDescriptor, is subject to change, and is already documented in bqSanitizedRandomUUIDForDescriptor

Reply:
Fixed.

Reply:
thanks! you can resolve this convo